Backend Comments Test Cases

misherwa opened the pull request:
*Please complete the checklist so the group can quickly understand your PR.*

### Category
- [x] **Feature**
- [ ] **Bug Fix**
- [x] **Testing**
- [ ] **Documentation**
- [ ] **Other**

### Domain
- [ ] **Frontend**
- [x] **Backend**
- [ ] **Both**

### Description
Provide test case coverage for all the files in the Comments backend folder. I have completed test coverage for models.py, serializers.py, and views, py. I've also added error handling and authorization requirements features. In views.py, a look_up field was added to retrieve a certain comment, and deleting can be done by the author, moderator, or hub owner.

### Related Issue(s)
#44 "Backend Comments Need Test Cases"

### New Additions
No new files

### Modifications
-tests.py (Added test coverage for models.py)
-urls.py (Removed a copy line for commentdelete)
-serializers.py (Added file changes from PR #67 and error handling)
-views.py (Added authorization and error handling)

### New Dependencies
None

### Visuals
![q4](https://github.com/user-attachments/assets/5041e868-c50b-48f0-b78b-b7042edb58ba)
Here is a current view of the coverage results for the Comments folder with 93 percent coverage. 

### Architecture of Your Changes
Made authorization checks in views.py for the CommentDelete class. This is to ensure that another user can not delete a comment. Authorization checks for replying to liking or disliking replies, which should not be possible since the feature has not been implemented to replies. Liking and disliking are currently only for comments or posts.

No other modifications to an existing system, only improvements to test coverage results and error handling.

### Usage and Integration
Just used to increase our overall test coverage percentage,

### Testing
Describe the tests performed to ensure your changes function as expected. This section should include:
- **Testing Approach**: Applied unit testing 
- **Edge Cases**: Currently need to add coverage for test cases that involve the serializers.py and views.py files in the Comments backend folder.
- **Test Commands**: Make sure to do "pip install coverage", then in the backend folder do "python manage.py test". Lastly, you can do "coverage report" for the result

### Mentions
N/A

thecatfour commented:
I don't think "test_create_comment" is the best name for this test case. The name implies that the test is creating a comment, but the comment is created in the setUp function. Something like "test_check_setup_comment" might be more accurate.

![image](https://github.com/user-attachments/assets/1eeaa28c-5a32-477b-b180-573a15b7f0e4)

leed85 reviewed (commented):
Good tests, I think we need to discuss the architecture of comments a bit, should we allow constant reply of comments(i.e. reddit style where you can reply to a specific comment or a reply and it constantly branches off from there) or like youtube where you just @ the person in the comments as a reply

misherwa commented:
> test_check_setup_comment

Yeah that makes sense, I changed "test_create_comment" to "test_check_setup_comment"

misherwa commented:
> Good tests, I think we need to discuss the architecture of comments a bit, should we allow constant reply of comments(i.e. reddit style where you can reply to a specific comment or a reply and it constantly branches off from there) or like youtube where you just @ the person in the comments as a reply

I believe it should be more like YouTube. It's more of a personal preference for me, but I think it would be safer to implement for now and we can try testing out a branching method,

misherwa commented:
I modified the "to_representation" and "validate" methods in the  seralizers.py file. I added error handling for both. 

The error handling for "to_representation" can be found in the if statement right before "representation" gets returned. It is also in the CommentSerializer class.

"validate" is in the "CommentCreateSerializer" class, which includes "try and except" when issues if a comment or post doesn't exist

thecatfour commented:
Is there a test for checking if a hub owner/mod can delete a comment on a post in their hub? I can only find 2 delete tests that involve a non-author trying to delete a comment and the author trying to delete a comment. I think its important to make sure mods and owners can delete comments, so I think you should add those cases.

misherwa commented:
> Is there a test for checking if a hub owner/mod can delete a comment on a post in their hub? I can only find 2 delete tests that involve a non-author trying to delete a comment and the author trying to delete a comment. I think its important to make sure mods and owners can delete comments, so I think you should add those cases.

Gotcha, I updated the code to include test cases for when a hub owner or mod wants to delete a comment:
![org](https://github.com/user-attachments/assets/806688a3-e7a8-4bb0-a6e9-8a2614561930)


Note: I had to update the views.py file for Comments to include cases if a hub owner or mod wants to delete a comment. I also added a look_up field and set it to "comment_id".  It's similar  to how it's done in for views.py in Posts, so now the "CommentDelete" class looks like this:
![cur](https://github.com/user-attachments/assets/1b552c07-e550-480d-b684-d0e34f805a19)

thecatfour commented:
I'm not sure if including the .coverage file in the PR is necessary. It might cause conflicts similar to the ones we got earlier when trying to merge with the database file. Other than that, I think this includes all the tests that we need for the comments right now. If we include a view for editing a comment, we will need to add test cases for that though.

nnguyen144 commented:
> > Good tests, I think we need to discuss the architecture of comments a bit, should we allow constant reply of comments(i.e. reddit style where you can reply to a specific comment or a reply and it constantly branches off from there) or like youtube where you just @ the person in the comments as a reply
> 
> I believe it should be more like YouTube. It's more of a personal preference for me, but I think it would be safer to implement for now and we can try testing out a branching method,

YouTube comment styling is easier and just requires a link to the post and to another comment if the comment is a reply. I believe YouTube comment styling would be the way to go.

nnguyen144 commented on backend/Comments/tests.py:
```diff
@@ -1,3 +1,399 @@
+    
+    # Create a different user who is neither the author and should not have permission to delete  
+    def test_unauthorized_comment_delete_permission(self):
+        other_user = User.objects.create_user(username='other_user', password='other_user_passwird')
```
Should be 'other_user_password' and not 'other_user_passwird'

nnguyen144 commented on backend/Comments/tests.py:
```diff
@@ -1,3 +1,399 @@
+        response = self.client.delete(url)
+        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN)
+        
+    def test_Moderator_deleting_comment(self):
```
Small nitpick but you could probably de-capitalize Moderator and Hub in the titles of these two tests

nnguyen144 reviewed (changes requested):
I also noticed that you reversed a lot of urls that are not in urls.py, such as 
url = reverse("comments:like-comment", kwargs={"comment_id": reply.id})
Make sure you fix this.

nnguyen144 commented:
Also since you added a new URL, can you add that URl to url-segments.ts

thecatfour commented:
> Also since you added a new URL, can you add that URl to url-segments.ts

I already implemented the new URLs in #67, so it should be the same whether they are added here or in that PR.

thecatfour reviewed (approved):
I believe this addresses what we need for the backend comments tests.

misherwa commented:
> I believe this addresses what we need for the backend comments tests.

I think I'll wait until your #67 PR gets merged before making proper changes to this one. But for now, I'll wait until your PR is ready to be reviewed and I can take a look at it. Or I can take a look at your draft if now if you think I should.
